### PR TITLE
Test coverage

### PR DESCRIPTION
**Pull request by SkillfulDev**
image
image
image
image
image
image
image

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,9 @@
+MYSQLDB_USER=root
+MYSQLDB_ROOT_PASSWORD=pass
```
Comment (k0sm0naft) [inline .env]: Better doesn't show credential, and make `.envSample` for git.

File: pom.xml [hunk 2]
```diff
@@ -22,6 +22,7 @@
         <liquibase.version>4.22.0</liquibase.version>
         <maven.checkstyle.plugin.configLocation>checkstyle.xml</maven.checkstyle.plugin.configLocation>
         <jjwt.version>0.11.5</jjwt.version>
+        <testcontainers.version>1.19.7</testcontainers.version>
```
Comment (k0sm0naft) [inline pom.xml:25]: Just advice: you can use version from parent pom.

File: pom.xml [hunk 3]
```diff
@@ -32,6 +33,10 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-data-jpa</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.springframework.security</groupId>
+            <artifactId>spring-security-test</artifactId>
+        </dependency>
         <dependency>
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-test</artifactId>
```
Comment (k0sm0naft) [inline pom.xml:42]: Better remove dependency that you don't use:
        <dependency>
            <groupId>com.h2database</groupId>
            <artifactId>h2</artifactId>
            <scope>test</scope>
        </dependency>

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/CategoryServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        categoryDto.setName(SPECIFIC_CATEGORY_NAME);
+
+        when(categoryRepository.findById(categoryId)).thenReturn(java.util.Optional.of(category));
+        when(categoryMapper.toDto(category)).thenReturn(categoryDto);
+
+        CategoryResponseDto result = categoryService.getById(categoryId);
+
+        assertEquals(SPECIFIC_CATEGORY_NAME, result.getName());
+    }
+
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/CategoryServiceImplTest.java]:
```suggestion
```

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/CategoryServiceImplTest.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
+package ua.chernonog.onlinebookstore.service.impl;
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/CategoryServiceImplTest.java:1]: You can add check for count of method calls.

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 6]
```diff
@@ -0,0 +1,206 @@
+package ua.chernonog.onlinebookstore.service.impl;
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java:1]: You can add check for count of method calls.

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+        responseDto.setTitle(SET_BOOK_TITLE);
+
+        Pageable pageable = PageRequest.of(DEFAULT_PAGE_NUMBER, DEFAULT_PAGE_SIZE);
+
+        Mockito.when(bookRepository.findAll(pageable)).thenReturn(page);
+        Mockito.when(bookMapper.toDto(book)).thenReturn(responseDto);
+
+        List<BookResponseDto> books = bookService.getAll(pageable);
+
+        assertEquals(1, books.size());
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java]:
```suggestion
        assertEquals(page.getSize(), books.size());
```
And the same behavior in other test with collection.

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+
+        assertEquals(UPDATED_BOOK_TITLE, updatedBook.getTitle());
+        assertEquals(UPDATED_BOOK_AUTHOR, updatedBook.getAuthor());
+        assertEquals(UPDATED_BOOK_COVER_IMAGE, updatedBook.getCoverImage());
+    }
+
+    @Test
+    @DisplayName("Delete a book by ID")
+    void deleteById_WithValidId_ShouldDeleteTheBook() {
+        Long bookId = 1L;
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java]: You can make it constant.

File: src/test/java/ua/chernonog/onlinebookstore/controller/CategoryControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import org.springframework.test.web.servlet.MockMvc;
+import org.springframework.test.web.servlet.setup.MockMvcBuilders;
+import org.springframework.web.context.WebApplicationContext;
+import ua.chernonog.onlinebookstore.dto.request.category.CreateCategoryDto;
+import ua.chernonog.onlinebookstore.dto.response.book.BookDtoWithoutCategoryIds;
+import ua.chernonog.onlinebookstore.dto.response.category.CategoryResponseDto;
+import ua.chernonog.onlinebookstore.service.CategoryService;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+class CategoryControllerTest {
```
Comment (k0sm0naft) [inline src/test/java/ua/chernonog/onlinebookstore/controller/CategoryControllerTest.java:42]: For me - better do not mock service layer and make controller tests - integration tests.

**Review by k0sm0naft — CHANGES_REQUESTED** [review]
Needed some improvements.

File: src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        responseDto.setTitle(bookDto.getTitle());
+        responseDto.setAuthor(bookDto.getAuthor());
+        responseDto.setIsbn(bookDto.getIsbn());
+        responseDto.setPrice(bookDto.getPrice());
+        responseDto.setDescription(bookDto.getDescription());
+        responseDto.setCoverImage(bookDto.getCoverImage());
+
+        when(bookService.save(any(CreateBookRequestDto.class))).thenReturn(responseDto);
+
+        objectMapper = new ObjectMapper();
```
Comment (andrii-hoienko) [inline src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java]: What is the reason to create object mapper in most methods?

File: src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java [hunk 11]
```diff
@@ -0,0 +1,165 @@
...
+        String bookJson = objectMapper.writeValueAsString(bookDto);
+
+        mockMvc.perform(post(BOOKS_BASE_URL)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(bookJson))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.title").value(TITLE));
```
Comment (andrii-hoienko) [inline src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java]:
```suggestion
                .andExpect(jsonPath(TITLE_JSON_PATH).value(TITLE)); //or something like that
```

File: src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java [hunk 12]
```diff
@@ -0,0 +1,165 @@
...
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    @WithMockUser(roles = "USER")
+    @DisplayName("Test book deletion fails for non-admin user")
+    void deleteBook_WithUserRole_ShouldFail() throws Exception {
+        Long bookId = 1L;
+
+        mockMvc.perform(delete("/books/{id}", bookId))
```
Comment (andrii-hoienko) [inline src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java]: replace literal with constant

File: src/test/java/ua/chernonog/onlinebookstore/controller/CategoryControllerTest.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        responseDto.setDescription(createCategoryDto.getDescription());
+
+        when(categoryService.save(any(CreateCategoryDto.class))).thenReturn(responseDto);
+
+        mockMvc.perform(post(CATEGORIES_BASE_URL)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(objectMapper.writeValueAsString(createCategoryDto))
+                        .with(csrf()))
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.name").value("Science Fiction"));
```
Comment (andrii-hoienko) [inline src/test/java/ua/chernonog/onlinebookstore/controller/CategoryControllerTest.java]: Why sometimes you use constants for jsonPath and sometimes not? Please be consistent and choose approach

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+        requestDto.setDescription(NEW_BOOK_DESCRIPTION);
+        requestDto.setCoverImage(NEW_BOOK_COVER_IMAGE);
+
+        BookResponseDto responseDto = new BookResponseDto();
+        responseDto.setTitle(requestDto.getTitle());
+        responseDto.setAuthor(requestDto.getAuthor());
+        responseDto.setIsbn(requestDto.getIsbn());
+        responseDto.setPrice(requestDto.getPrice());
+        responseDto.setDescription(requestDto.getDescription());
+        responseDto.setCoverImage(requestDto.getCoverImage());
```
Comment (andrii-hoienko) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java]: Code duplication

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java (hunk 7 above)
Comment (andrii-hoienko) [reply src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java]: Not resolved

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Very nice, left some comments

File: src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java (hunk 10 above)
Comment (SkillfulDev) [reply src/test/java/ua/chernonog/onlinebookstore/controller/BookControllerTest.java]: There is no any reason. I forgot to remove.

File: src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+        Long bookId = VALID_BOOK_ID;
+        Book existingBook = new Book();
+        existingBook.setId(bookId);
+        existingBook.setTitle(OLD_BOOK_TITLE);
+
+        BookResponseDto responseDto = getBookResponseDto(requestDto);
+
+        Mockito.when(bookRepository.findById(bookId)).thenReturn(Optional.of(existingBook));
+        Mockito.when(bookRepository.save(existingBook)).thenReturn(existingBook);
+        Mockito.when(bookMapper.toDto(existingBook)).thenReturn(responseDto);
```
Comment (yevheniisynytsia) [inline src/test/java/ua/chernonog/onlinebookstore/service/impl/BookServiceImplTest.java:154]: make static import

**Review by yevheniisynytsia — APPROVED** [review]

**Pull request merged**